core/shared/src/main/scala/zio/internal/UpdateOrderLinkedMap.scala

```diff
@@ -0,0 +1,232 @@
+package zio.internal
+
+import scala.annotation.tailrec
+import scala.collection.AbstractIterator
+import scala.collection.immutable.{HashMap, VectorBuilder}
+import scala.util.hashing.MurmurHash3
+
+final class UpdateOrderLinkedMap[K, +V](
+  fields: Vector[Any],
+  underlying: HashMap[K, (Int, V)]
+) { self =>
+  import UpdateOrderLinkedMap._
+
+  def size: Int = underlying.size
+
+  def isEmpty: Boolean = size == 0
+
+  def updated[V1 >: V](key: K, value: V1): UpdateOrderLinkedMap[K, V1] = {
```

Comment:
TODO: Need to add a hard limit on how big the underlying can get before rebuilding it fresh. It's very unlikely we'll ever have a case that keeps forever adding services but it's not impossible